mlflow/anthropic/autolog.py

```diff
@@ -36,6 +41,25 @@ def patched_class_call(original, self, *args, **kwargs):
         ) as span:
             inputs = construct_full_inputs(original, self, *args, **kwargs)
             span.set_inputs(inputs)
-            outputs = original(self, *args, **kwargs)
-            span.set_outputs(outputs)
+
+            messages = [convert_message_to_mlflow_chat(msg) for msg in inputs.get("messages", [])]
+            try:
+                outputs = original(self, *args, **kwargs)
+                span.set_outputs(outputs)
+            finally:
+                # Set message attribute once at the end to avoid multiple JSON serialization
+                try:
+                    messages.append(convert_message_to_mlflow_chat(outputs))
+                    set_span_chat_messages(span, messages)
+                except Exception as e:
+                    _logger.debug(f"Failed to set chat messages for {span}. Error: {e}")
+                    raise
+
+                try:
```

Comment:
do we need to have this logic inside the "finally" block?

Reply:
Yes, even when an exception is raised, we should log messages and tools attributes.

Reply:
Oh wait, if you are talking about tool attribute specifically, we can set it before calling the model. But maybe having two in the same place is better for readability.

Reply:
I was talking about the tool attributes. I don't think having tool attribute within the try/finally block of the original function is necessary, and we should minimize the nesting of try/catch if possible.

Reply:
> we should minimize the nesting of try/catch if possible. 

I see your point, but what I'm saying is, message attribute will still within finally block 
so it doesn't matter about "nesting", and moving the other one outside makes code more messy.

Reply:
>message attribute will still within the finally block so it doesn't matter about "nesting"

Generally, messages should be distinct from tool definition logging and we can log tool definitions regardless of the status of model calling or message extraction. Also, with the current code, tool defs are logged after message logging. So if message logging fails, users cannot see tool defs either.

>moving the other one outside makes code more messy

I understand you want to have one place for "attribute logging". But, it is normal to have pre-execution logging and post-execution logging separately. For Authropic specifically, input and tool def are pre-execution logging and output and chat messages are post-execution. Since you can put tool def together with input logging, I don't think it'll be messy.

Reply:
ok ok I back down here, the discussion became a bit philosophical😅

Btw maybe it's worthwhile to document some style policy for the team. We use Google coding style but it is not very detailed. Ideally we don't spend much time on the styling during code review.

Reply:
+1, also styling comments should be nit and non-blocking, so it's mostly up to code author